chronos/asyncproc.nim

```diff
@@ -0,0 +1,1323 @@
+    else:
+      err(osLastError())
+
+  proc terminate(p: AsyncProcessRef): AsyncProcessResult[void] =
+    if terminateProcess(p.processHandle, 0) != 0'u32:
+      ok()
+    else:
+      err(osLastError())
+
+  proc kill(p: AsyncProcessRef): AsyncProcessResult[void] =
+    p.terminate()
+
+  proc running(p: AsyncProcessRef): AsyncProcessResult[bool] =
+    let res = ? p.peekExitCode()
+    if res == -1:
+      ok(true)
+    else:
+      ok(false)
+
+  proc waitForExit*(p: AsyncProcessRef,
+                    timeout = InfiniteDuration): Future[int] {.async.} =
+    if p.exitStatus.isSome():
+      return p.exitStatus.get()
+
+    let wres =
+      try:
+        await waitForSingleObject(p.processHandle, timeout)
+      except ValueError as exc:
+        raiseAsyncProcessError("Unable to wait for process handle", exc)
+
+    if wres == WaitableResult.Timeout:
+      let res = p.kill()
+      if res.isErr():
+        raiseAsyncProcessError("Unable to terminate process", res.error())
+
+    let exitCode = p.peekProcessExitCode().valueOr:
+      raiseAsyncProcessError("Unable to peek process exit code", error)
+
+    if exitCode >= 0:
+      p.exitStatus = some(exitCode)
+    return exitCode
+
+  proc peekExitCode(p: AsyncProcessRef): AsyncProcessResult[int] =
+    if p.exitStatus.isSome():
+      return ok(p.exitStatus.get())
+    let res = waitForSingleObject(p.processHandle, DWORD(0))
+    if res != WAIT_TIMEOUT:
+      let exitCode = ? p.peekProcessExitCode()
+      ok(exitCode)
+    else:
+      ok(-1)
+else:
+  import std/strutils
+
+  type
+    SpawnAttr = object
+      attrs: PosixSpawnAttr
+      actions: PosixSpawnFileActions
+
+  proc fd(h: AsyncStreamHolder): cint =
+    case h.kind
+    of StreamKind.Reader:
+      cint(h.reader.tsource.fd)
+    of StreamKind.Writer:
+      cint(h.writer.tsource.fd)
+    of StreamKind.None:
+      raiseAssert "Incorrect stream holder"
+
+  proc isEmpty(h: AsyncStreamHolder): bool =
+    h.kind == StreamKind.None
+
+  proc initSpawn(pipes: AsyncProcessPipes, options: set[AsyncProcessOption]
+                ): Result[SpawnAttr, OSErrorCode] =
+    template doCheck(body: untyped): untyped =
+      let res = body
+      if res != 0:
+        return err(OSErrorCode(res))
+
+    var
+      attrs =
+        block:
+          var value: PosixSpawnAttr
+          let res = posixSpawnAttrInit(value)
+          if res != 0:
+            return err(OSErrorCode(res))
+          value
+      actions =
+        block:
+          var value: PosixSpawnFileActions
+          let res = posixSpawnFileActionsInit(value)
+          if res != 0:
+            discard posixSpawnAttrDestroy(attrs)
+            return err(OSErrorCode(res))
+          value
+      mask =
+        block:
+          var res: Sigset
+          discard sigemptyset(res)
+          res
+
+    doCheck(posixSpawnAttrSetSigMask(attrs, mask))
+    if AsyncProcessOption.ProcessGroup in options:
+      doCheck(posixSpawnAttrSetPgroup(attrs, 0))
```

Comment:
What is this option trying to do? It sets the `pgid` of the created process to zero? Thus, allowing only a single process group to exist in the application?

I think the right API is to create a `ProcessGroup` object before spawning the process and to pass it as an input to `startProcess`. This would allow me to create separate process groups which I can wait for completion or terminate at will (as groups).

On Windows, I expect the same API to be mapped to [Job Objects](https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects?redirectedfrom=MSDN).

Reply:
I think the most useful aspect of this "Process Group" API is that it can help you kill the entire subtree of processes launched by a particular program that you are starting. By default, such indirect child processes will be orphaned on Posix, which is painful for the users.

The job objects on Windows offer the same functionality, with even stronger guarantees.

Reply:
A related common problem is this one:
https://stackoverflow.com/questions/284325/how-to-make-child-process-die-after-parent-exits

... but I'll grant you that there aren't very good cross-platform solutions that have precisely the same semantics everywhere.